src/mouse_events.cpp

```diff
@@ -978,9 +978,9 @@ void mouse_handler::move_action(bool browse)
 					}
 
 					// block where we temporary move the unit
-					{
-						temporary_unit_mover temp_mover(pc_.get_units(), src, attack_from, itor->move_left, true);
-						choice = show_attack_dialog(attack_from, clicked_u->get_location());
+					{//fork 
+						//temporary_unit_mover temp_mover(pc_.get_units(), src, attack_from, itor->move_left, true);//this is used to move unit to `attackfrom`.
+						choice = show_attack_dialog(src, clicked_u->get_location(), false);
```

Comment:
This seems like it's going to give the wrong answer in some cases. Consider a scenario with the following conditions:

1. Red elvish archer is standing on mountains.
2. Blue elvish archer is standing on forest.
3. North of blue elf is forest.
4. South of blue elf is flat.
5. It's red's turn.

Now, there are two possibilities here (ignoring the diagonal directions to keep things simple):

1. Red attacks blue from the north. It has a 30% chance to be hit.
2. Red attacks blue from the south. It has a 60% chance to be hit.

The 30% or 60% chance _should_ be shown in the attack dialog, depending on which of the two adjecent hexes was hovered before the mouse was moved over the blue elf. But with this change, it looks like the attack dialog will instead show a 40% chance to be hit. Could you test and see if this is the case?

I think _both_ locations probably need to be passed to show_attack_dialog – that is, both `src` and `attack_from`.

Reply:
Also ability calculations (leadership for example) depend on the hex that the unit is standing on, and there are some that affect the hexes around the unit that's moving - for example illuminates. That needs the correct hex to be passed around not just in the ability calculation code, but also into the time of day code.

The ability calculation code is already sufficiently complex that we're having trouble handling changes to it.